src/wast-parser.cc

```diff
@@ -937,10 +942,10 @@ Result WastParser::ParseValueTypeList(TypeVector* out_type_list,
     CHECK_RESULT(ParseValueType(&type));
 
     if (type.is_index()) {
-      // TODO: Incorrect values can be misinterpreted by the parser.
-      if (type.index() >= static_cast<Index>(Type::Void)) {
+      if (IsIndexLikelyType(type.index())) {
```

Comment:
From the code that follows it looks like we are trying to distinguish reference types from function types?  Is that right?  How about something like `IsReferenceType`?   I'm not really sure what `IsIndexLikelyType` means here... perhaps I'm misunderstanding?

Reply:
It is not just reference type, it is currently any type. The `ParseValueType` encodes the value type as index.

For example: the binary code of `(func $f (param i64 (ref 0xfffffffe)))` is

```
; func type 0
000000b: 60                                        ; func
000000c: 02                                        ; num params
000000d: 7e                                        ; i64
000000e: 7e                                        ; i64
000000f: 00                                        ; num results
```

The i64 is encoded as `0xfffffffe`, which is stored in the index. Also, the `(ref 0xfffffffe)` is converted to i64, because currently it is impossible to tell the difference (at the moment). The `function_ref` patch will fix this issue.

Reply:
SGTM then.